src/cpp/session/modules/SessionUserPrefs.cpp

```diff
@@ -0,0 +1,262 @@
+   if (error)
+   {
+      r::exec::error("Could not save preferences: " + error.description());
+      return false;
+   }
+
+   // let other modules know we've updated the prefs
+   module_context::events().onPreferencesSaved();
+   
+   return true;
+}
+
+SEXP rs_removePref(SEXP prefName)
+{
+   // extract name of preference to remove
+   std::string pref = r::sexp::safeAsString(prefName, "");
+   if (pref.empty())
+      return R_NilValue;
+   
+   Error error = userPrefs().clearValue(pref);
+   if (error)
+   {
+      r::exec::error("Could not save preferences: " + error.description());
+   }
+
+   module_context::events().onPreferencesSaved();
+
+   return R_NilValue;
+}
+
+SEXP rs_readPref(Preferences& prefs, SEXP prefName)
+{
+   r::sexp::Protect protect;
+
+   // extract name of preference to read
+   std::string pref = r::sexp::safeAsString(prefName, "");
+   if (pref.empty())
+      return R_NilValue;
+
+   auto prefValue = prefs.readValue(pref);
+   if (prefValue)
+   {
+      // convert to SEXP and return
+      return r::sexp::create(*prefValue, &protect);
+   }
+
+   // No preference found with this name
+   return R_NilValue;
+}
+
+SEXP rs_readUserPref(SEXP prefName)
+{
+   return rs_readPref(userPrefs(), prefName);
+}
+
+SEXP rs_writeUserPref(SEXP prefName, SEXP value)
+{
+   if (writePref(userPrefs(), prefName, value))
+   {
+      userPrefs().notifyClient(kUserPrefsUserLayer, r::sexp::safeAsString(prefName));
+   }
+
+   return R_NilValue;
+}
+
+SEXP rs_readUserState(SEXP stateName)
+{
+   return rs_readPref(userState(), stateName);
+}
+
+SEXP rs_writeUserState(SEXP stateName, SEXP value)
+{
+   if (writePref(userState(), stateName, value))
+   {
+      userState().notifyClient(kUserStateUserLayer, r::sexp::safeAsString(stateName));
+   }
+
+   return R_NilValue;
+}
+
+SEXP rs_allPrefs()
+{
+   r::sexp::Protect protect;
+
+   // Create table of all preferences
+   std::vector<std::string> keys = userPrefs().allKeys();
+   std::vector<std::string> sources;
+   std::vector<std::string> values;
+   for (const auto key: keys) 
+   {
+      std::string layer;
+      auto val = userPrefs().readValue(key, &layer);
+      if (val)
+      {
+         sources.push_back(layer);
+
+         std::ostringstream oss;
+         json::write(*val, oss);
+         values.push_back(oss.str());
+      }
+      else
+      {
+         sources.push_back("none");
+         values.push_back("");
+      }
+   }
+   
+   // Convert to data frame
+   std::vector<std::string> names({"Preference", "Source", "Value"});
+   SEXP list = r::sexp::createList(names, &protect);
+   SET_VECTOR_ELT(list, 0, r::sexp::create(keys, &protect));
+   SET_VECTOR_ELT(list, 1, r::sexp::create(sources, &protect));
+   SET_VECTOR_ELT(list, 2, r::sexp::create(values, &protect));
+   r::exec::RFunction asDataFrame("as.data.frame");
+   asDataFrame.addParam("x", list);
+   asDataFrame.addParam("stringsAsFactors", false);
+   SEXP frame = R_NilValue;
```

Comment:
Does this work with `NULL`? I'm a bit surprised (would expect this parameter needs to be an environment, e.g. R_BaseEnv). Or perhaps we remap `R_NilValue` to some environment?

Either way, I believe it would be better to be explicit.

Reply:
Which parameter are you referring to? `frame` here is just an SEXP that will hold the result of `as.data.frame` (I don't think its initial value matters and am just zero-initializing here)

Reply:
Sorry, you're right -- my editor had brought me to the wrong overload of `RFunction::call()` and so I misunderstood what was happening here.

(not sure if it's worth worrying about, but this does imply that `as.data.frame()` will be discovered in the R global environment, which could present issues if the user has overridden that function or provided their own definition)